src/audio/kpb.c

```diff
@@ -2378,22 +2398,202 @@ static int kpb_set_micselect(struct comp_dev *dev, const void *data,
 	return 0;
 }
 
+static int devicelist_push(struct device_list *devlist, struct comp_dev **dev)
+{
+	if (devlist->count != DEVICE_LIST_SIZE) {
+		devlist->devs[devlist->count] = dev;
+		devlist->count++;
+		return 0;
+	}
+	return -EINVAL;
+}
+
+static void devicelist_reset(struct device_list *devlist, bool remove_items)
+{
+	/* clear items */
+	if (remove_items) {
+		for (int i = 0; i < DEVICE_LIST_SIZE; i++)
+			*devlist->devs[i] = NULL;
+	}
+	/* zero the pointers */
+	for (int i = 0; i < DEVICE_LIST_SIZE; i++)
+		devlist->devs[i] = NULL;
+
+	devlist->count = 0;
+}
+
+static inline int alloc_fmt_module_list_item(struct kpb_fmt_dev_list *fmt_device_list,
+					     struct comp_dev *mi_ptr, struct comp_dev ***item)
+{
+	/* -1 means we did not find the slot yet */
+	int first_empty_slot_idx = -1;
+
+	for (size_t module_slot_idx = 0; module_slot_idx < FAST_MODE_TASK_MAX_MODULES_COUNT;
+	     ++module_slot_idx){
+		/* check if module already added */
+		if (fmt_device_list->modules_list_item[module_slot_idx] == mi_ptr)
+			return -EINVAL;
+		/* finding first available empty slot */
+		if (first_empty_slot_idx < 0 &&
+		    !fmt_device_list->modules_list_item[module_slot_idx])
+			first_empty_slot_idx = module_slot_idx;
+	}
+	/* add item to first available empty slot */
+	if (first_empty_slot_idx >= 0) {
+		fmt_device_list->modules_list_item[first_empty_slot_idx] = mi_ptr;
+		*item = &fmt_device_list->modules_list_item[first_empty_slot_idx];
+		return 0;
+	}
+	return -ENOMEM;
+}
+
+static int prepare_fmt_modules_list(struct comp_dev *kpb_dev,
+				    uint32_t outpin_idx,
+				    const struct kpb_task_params *modules_to_prepare)
+{
+	if (!kpb_dev)
+		return -EINVAL;
+	if (!modules_to_prepare)
+		return -EINVAL;
+	if (outpin_idx >= KPB_MAX_SINK_CNT)
+		return -EINVAL;
+
+	int ret = 0;
+	struct comp_dev *dev = NULL;
+	struct kpb_fmt_dev_list *fmt_device_list =
+		&((struct comp_data *)comp_get_drvdata(kpb_dev))->fmt_device_list;
+
+	if (modules_to_prepare->number_of_modules != 0) {
+		fmt_device_list->kpb_list_item[outpin_idx] = kpb_dev;
+		ret = devicelist_push(&fmt_device_list->device_list[outpin_idx],
+				      &fmt_device_list->kpb_list_item[outpin_idx]);
+		if (ret < 0)
+			return ret;
+	}
+	for (size_t mod_idx = 0; mod_idx < modules_to_prepare->number_of_modules; ++mod_idx) {
+#ifdef CONFIG_IPC_MAJOR_4
+		uint32_t comp_id = IPC4_COMP_ID(modules_to_prepare->dev_ids[mod_idx].module_id,
+						modules_to_prepare->dev_ids[mod_idx].instance_id);
+
+		dev = ipc4_get_comp_dev(comp_id);
+#endif
+
+		if (!dev)
+			return -EINVAL;
```

Comment:
so for IPC3 this function will return `-EINVAL`, can we just make that explicit somehow?

Reply:
Yes, I do not support IPC3 here yet.  About initialisation of dev and the check - I could put those into the #ifdef, and if IPC4 is not defined just return in #else clause . Also I could add comment saying that  IPC3 support is not here yet. Does it sound good?

Reply:
IPC3 isn't supported at all by this, right? So I'd really put something like
```
if (IS_ENABLED(CONFIG_IPC_MAJOR_3))
    return -EINVAL;
```
somewhere very early, into some preparation function without which the whole fast mode task won't work

Reply:
Added this.
I'm also considering putting my new kpb.c functions in `#ifdef CONFIG_IPC_MAJOR_4` so that it's more explicit (instead of just this part).

Reply:
> I'm also considering putting my new kpb.c functions in `#ifdef CONFIG_IPC_MAJOR_4` so that it's more explicit (instead of just this part).

I think yes, that would be the best for now. You do need a preprocessor check there because of calling an IPC4 specific function, so just using `IS_ENABLED(CONFIG_IPC_MAJOR_3)` unfortunately isn't enough - that wasn't a very good idea, sorry. So, yes, I think putting everything under `#if IPC4` would be the best option ATM